src/option_learning.py

```diff
@@ -212,3 +216,226 @@ def update_segment_from_option_spec(self, segment: Segment,
+                changing_objects = [inv_sub[v] for v in changing_parameters]
+                initial_state = segment.states[0]
+                final_state = segment.states[-1]
+                absolute_params = final_state.vec(changing_objects)
+                option_param = absolute_params - initial_state.vec(
+                    changing_objects)
+
+                # Store the option parameterization for this segment so we can
+                # use it in update_segment_from_option_spec.
+                self._segment_to_grounding[segment] = (all_objects_in_operator,
+                                                       option_param)
+                del option_param  # not used after this
+
+                # Next, create the input vectors from all object states named
+                # in the operator parameters (not just the changing ones).
+                # Note that each segment contributions multiple data points,
+                # one per action.
+                assert len(segment.states) == len(segment.actions) + 1
+                for state, action in zip(segment.states, segment.actions):
+                    state_features = state.vec(all_objects_in_operator)
+                    # Compute the relative goal vector this segment.
+                    relative_goal_vec = absolute_params - state.vec(
+                        changing_objects)
+                    # Add a bias term for regression.
+                    x = np.hstack(([1.0], state_features, relative_goal_vec))
+                    X_regressor.append(x)
+                    Y_regressor.append(action.arr)
+
+            X_arr_regressor = np.array(X_regressor, dtype=np.float32)
+            Y_arr_regressor = np.array(Y_regressor, dtype=np.float32)
+            regressor = MLPRegressor()
+            print(f"Fitting regressor with X shape: {X_arr_regressor.shape}, "
+                  f"Y shape: {Y_arr_regressor.shape}.")
+            regressor.fit(X_arr_regressor, Y_arr_regressor)
+
+            # Construct the ParameterizedOption for this operator.
+            name = f"{op.name}LearnedOption"
+            parameterized_option = _LearnedNeuralParameterizedOption(
+                name, op, regressor, changing_parameters)
+            option_specs.append((parameterized_option, list(op.parameters)))
+
+        return option_specs
+
+    @staticmethod
+    def _get_changing_parameters(datastore: Datastore) -> Set[Variable]:
+        all_changing_variables = set()
+        for segment, sub in datastore:
+            start = segment.states[0]
+            end = segment.states[-1]
+            for o, v in sub.items():
+                if not np.array_equal(start[o], end[o]):
+                    all_changing_variables.add(v)
+        return all_changing_variables
+
+    def update_segment_from_option_spec(self, segment: Segment,
+                                        option_spec: OptionSpec) -> None:
```

Comment:
small thing, but it'd be nice to have a quick docstring here to explain what exactly this is doing

Reply:
This is overriding a parent class method. See docstring in https://github.com/Learning-and-Intelligent-Systems/predicators/blob/master/src/option_learning.py.

Reply:
ah didn't notice that; thanks for the pointer!